Sources/Build/BuildPlan.swift

```diff
@@ -307,19 +327,133 @@ public final class ClangTargetBuildDescription {
+                    )
+                } else {
+                    // When not building within a mixed target, use the custom
+                    // module map as it does not need to be modified.
+                    self.moduleMap = customModuleMapPath
+                }
             }
             // If a generated module map is needed, generate one now in our temporary directory.
             else if let generatedModuleMapType = clangTarget.moduleMapType.generatedModuleMapType {
-                let path = tempsPath.appending(component: moduleMapFilename)
-                let moduleMapGenerator = ModuleMapGenerator(targetName: clangTarget.name, moduleName: clangTarget.c99name, publicHeadersDir: clangTarget.includeDir, fileSystem: fileSystem)
-                try moduleMapGenerator.generateModuleMap(type: generatedModuleMapType, at: path)
-                self.moduleMap = path
+                let moduleMapPath = tempsPath
+                    .appending(component: "Product")
+                    .appending(component: moduleMapFilename)
+                try moduleMapGenerator.generateModuleMap(
+                    type: generatedModuleMapType,
+                    at: moduleMapPath,
+                    addSwiftSubmodule: isWithinMixedTarget
+                )
+
+                if isWithinMixedTarget {
+                    let intermediateModuleMapPath = tempsPath
+                        .appending(component: "Intermediates")
+                        .appending(component: moduleMapFilename)
+                    try moduleMapGenerator.generateModuleMap(
+                        type: .umbrellaDirectory(clangTarget.path),
+                        at: intermediateModuleMapPath,
+                        addSwiftSubmodule: true
+                    )
+
+                    // The underlying Clang target is building within a Mixed
+                    // language target and needs an auxiliary module map that
+                    // doesn't include the generated interop header from the
+                    // Swift half of the mixed target. This will later allow the
+                    // Clang half of the module to be built when compiling the
+                    // Swift part without the generated header being considered
+                    // an input (because it won't exist yet and is an output of
+                    // that compilation command).
+                    let unextendedModuleMapPath = tempsPath
+                        .appending(component: "Intermediates")
+                        .appending(component: unextendedModuleMapFilename)
+                    let nonObjcHeaders: [AbsolutePath] = clangTarget.headers
+                        .filter { $0.extension != "h" }
+                    try moduleMapGenerator.generateModuleMap(
+                        type: .umbrellaDirectory(clangTarget.path),
+                        at: unextendedModuleMapPath,
+                        excludeHeaders: nonObjcHeaders
```

Comment:
Also need to do this in the above block. Add a test and consolidate the two blocks since they are pretty much the same.

Reply:
Resolved via 8fa89f9 and 1a79ec9.